### PR TITLE
Tutorial for Tuner Subclassing

### PR DESCRIPTION
**Pull request by omalleyt12**

### CODE REVIEW

File: docs/templates/tutorials/subclass-tuner.md [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+        # natural concept of epoch, you do not have to call any of these hooks. In
+        # this case you should instead call `self.oracle.update_trial` and
+        # `self.oracle.save_model` manually.
+        for epoch in range(10):
+            print('Epoch: {}'.format(epoch))
+
+            self.on_epoch_begin(trial, model, epoch, logs={})
+            for batch, data in enumerate(train_ds):
+                self.on_batch_begin(trial, model, batch, logs={})
+                batch_loss = run_train_step(data).numpy()
```
Comment (fchollet) [inline docs/templates/tutorials/subclass-tuner.md]: This is a scalar so prefer doing `float(run_train_step(data))` (more readable than `.numpy()`, and gives you a plain Python float).

File: docs/templates/tutorials/subclass-tuner.md [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+                    labels, logits)
+                # Add any regularization losses.
+                if model.losses:
+                    loss += tf.math.add_n(model.losses)
+                gradients = tape.gradient(loss, model.trainable_variables)
+            optimizer.apply_gradients(zip(gradients, model.trainable_variables))
+            epoch_loss_metric.update_state(loss)
+            return loss
+
+        # `self.on_epoch_end` reports results to the `Oracle` and saves the
```
Comment (fchollet) [inline docs/templates/tutorials/subclass-tuner.md]: It should be specified in the text above ("Typically, `Tuner.run_trial` is the only method that users need to override when subclassing `Tuner`.") that `run_trial` should call `self.on_epoch_end` and optionally other callback methods such as `on_epoch_begin`, etc. The list of available methods, and which ones are optional or not, should be made explicit.

File: docs/templates/tutorials/subclass-tuner.md [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+The `Tuner` class at `kerastuner.engine.tuner.Tuner` can be subclassed to support advanced uses such as:
+
+- Custom training loops (GANs, reinforement learning, etc.)
+- Adding hyperparameters outside of the model builing function (preprocessing, data augmentation, test time augmentation, etc.)
+
+This tutorial will not cover subclassing to support non-Keras models. To accomplish this, you can subclass the `kerastuner.engine.base_tuner.BaseTuner` class (See `kerastuner.tuners.sklearn.Sklearn` for an example).
+
+### Understanding the search process.
+
+`Tuner.search` can be passed any arguments. These arguments will be passed directly to `Tuner.run_trial`, along with a `Trial` object that contains information about the current trial, including hyperparameters and the status of the trial. Typically, `Tuner.run_trial` is the only method that users need to override when subclassing `Tuner`.
```
Comment (fchollet) [inline docs/templates/tutorials/subclass-tuner.md]: Besides the detailed example below, I would recommend having a sketch of the code here, e.g.

```python

class MyTuner(kt.Tuner):

    def run_trial(self, trial, ...):
        hp = trial.hyperparameters
        model = self.hypermodel.build(trial.hyperparameters)
        for epoch in range(10):
              epoch_loss = ...
              self.on_epoch_end(trial, model, epoch, logs={'loss': epoch_loss})
```

Otherwise it's hard to parse what's essential and what's not.

File: docs/templates/tutorials/subclass-tuner.md [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+
+### Example
+
+```python
```
Comment (fchollet) [inline docs/templates/tutorials/subclass-tuner.md]: This is a complex example. We should consider using a format that enables us to make the example directly runnable in Colab in one click, and easily testable in an automatically generated integration test.

For instance, examples on tensorflow.org are typically shipped as runnable notebooks. The notebooks are used to generated a HTML webpage (it goes through markdown first), can be launched in Colab, and are run as part of integration tests.

File: docs/templates/tutorials/subclass-tuner.md (hunk 3 above)
Comment (omalleyt12) [reply docs/templates/tutorials/subclass-tuner.md]: Done

File: docs/templates/tutorials/subclass-tuner.md (hunk 4 above)
Comment (omalleyt12) [reply docs/templates/tutorials/subclass-tuner.md]: Is there a way to do the equivalent on GH? I couldn't find anything from googling around a bit

I ran this example on my computer locally to make sure it worked, is that ok for now?

File: docs/templates/tutorials/subclass-tuner.md (hunk 2 above)
Comment (omalleyt12) [reply docs/templates/tutorials/subclass-tuner.md]: Done

File: docs/templates/tutorials/subclass-tuner.md (hunk 1 above)
Comment (omalleyt12) [reply docs/templates/tutorials/subclass-tuner.md]: Done

**Comment by omalleyt12** [comment]
Thanks for the review!

File: docs/templates/tutorials/subclass-tuner.md (hunk 4 above)
Comment (fchollet) [reply docs/templates/tutorials/subclass-tuner.md]: There is no ready-made solution that I'm aware of, we'd have to roll out our own (and make it shared cross all of our repos). We'll do this eventually, but we can't block tutorials based on this, so let's ship the tutorial as is for now.

Comment (omalleyt12) [reply docs/templates/tutorials/subclass-tuner.md]: Sounds good, made a tracking bug here: https://github.com/keras-team/keras-autodoc/issues/35
